ddtrace/sampler.py

```diff
@@ -199,6 +200,34 @@ def __init__(
             )
             self.default_sampler = SamplingRule(sample_rate=default_sample_rate)
 
+    def parse_from_env_variable(self):
+        rules = []
+        # DD_TRACE_SAMPLING_RULES = ['{"sample_rate": "1", "service": "b", "name": 0.1}', ....]
+        env_rules = get_env("trace", "sampling_rules")
+        if env_rules is not None:
+            json_rules = []
+            try:
+                json_rules = json.loads(env_rules)
+            except json.decoder.JSONDecodeError:
+                log.debug(
+                    "Unable to set DD_TRACE_SAMPLING_RULES={%s} due to parsing configuration", env_rules, exc_info=True
+                )
+            # setting each rule from JSOB object
+            for rule in json_rules:
+                try:
+                    sample_rate = float(rule["sample_rate"])
+                    service = rule.get("service", SamplingRule.NO_RULE)
+                    name = rule.get("name", SamplingRule.NO_RULE)
+                    sampling_rule = SamplingRule(sample_rate=sample_rate, service=service, name=name)
+                except ValueError:
+                    log.debug("Invalid value while setting SamplingRule", exc_info=True)
+                    continue
```

Comment:
we should consider if we want to just skip this or actually raise an exception. since as-is it'll fail silently (unless they have debug logging enabled) and then the sampler will be misconfigured.

either letting these raise as-is (don't do any `except`) or raising new/better errors might be best? wdyt?

Reply:
True, failing silently isn't a great option as setting the sample rate could affect the ingestion of spans for a user. Would you rather I remove the except and have the application crash at startup?
```
~ DD_TRACE_SAMPLING_RULES='[{"sample_rate":2.0,"service":"abc","name":"abc"}]' ddtrace-run flask run                                                                                                           
Traceback (most recent call last):
  File "/usr/local/bin/ddtrace-run", line 33, in <module>
    sys.exit(load_entry_point('ddtrace', 'console_scripts', 'ddtrace-run')())
  File "/usr/local/bin/ddtrace-run", line 25, in importlib_load_entry_point
    return next(matches).load()
... 
  File "/Users/bowen.brooks/Documents/APM_Sandbox/dd-trace-py/ddtrace/__init__.py", line 14, in <module>
    tracer = Tracer()
  File "/Users/bowen.brooks/Documents/APM_Sandbox/dd-trace-py/ddtrace/tracer.py", line 128, in __init__
    self.sampler = DatadogSampler()  # type: BaseSampler
  File "/Users/bowen.brooks/Documents/APM_Sandbox/dd-trace-py/ddtrace/sampler.py", line 180, in __init__
    rules = self._parse_rules_from_env_variable()
  File "/Users/bowen.brooks/Documents/APM_Sandbox/dd-trace-py/ddtrace/sampler.py", line 221, in _parse_rules_from_env_variable
    sampling_rule = SamplingRule(sample_rate=sample_rate, service=service, name=name)
  File "/Users/bowen.brooks/Documents/APM_Sandbox/dd-trace-py/ddtrace/sampler.py", line 344, in __init__
    raise ValueError(
ValueError: SamplingRule(sample_rate={!r}) must be greater than or equal to 0.0 and less than or equal to 1.0
```
Or catch it and raise and error log so the application can still run but the tracer doesn't have to be in debug to notice the error.
```
except ValueError:
      log.error("Invalid value while setting SamplingRule", exc_info=True)
      continue
```

Reply:
I'm in favor of the second option as to prevent an application from crashing.

Reply:
I actually prefer the first option and letting the application crash.

The one thing we can do to improve this whole thing maybe having something like:

```python
def __init__(self, ...):
    ...

    if rules is None:
        env_rules = get_env("trace", "sampling_rules")
        if env_rules:
            try:
                rules = self._parse_from_env_variable(env_rules)
            except Exception as e:
                raise ValueError("Provided DD_TRACE_SAMPLING_RULES={0!r} were not valid: {1}".format(env_rules, e)
        else:
            rules = []

    ...
```

That is, let the helper raise whatever error is gets, but then we can do the catching and re-raise a more appropriate error from the `__init__` function.

Here I am catching just `Exception`, but we could catch the JSON decode error and give a better message like "provided rules are not valid json" vs "rule provided was invalid because you didn't give us a sample rate"... etc